Add basic-auth flag to ngrok

stasadev opened the pull request:
## The Issue

The documentation includes an [example](https://ddev.readthedocs.io/en/stable/users/usage/commands/#share) for ngrok:

```bash
# Share the current project using ngrok’s basic-auth argument
ddev share --basic-auth username:pass1234
```

The same example is provided by the `ddev help` command:
```bash
$ ddev help share
Use "ddev share" or add on extra ngrok commands, like "ddev share --basic-auth username:pass1234". Although a few ngrok commands are supported directly, any ngrok flag can be added in the ngrok_args section of .ddev/config.yaml. Requires a free or paid account on ngrok.com; use the "ngrok authtoken" command to set up ngrok.

Usage:
  ddev share [project] [flags]

Examples:
ddev share
ddev share --basic-auth username:pass1234
ddev share myproject

Flags:
  -h, --help               help for share
      --subdomain string   ngrok --subdomain argument, as in "ngrok --subdomain my-subdomain:, requires paid ngrok.com account"

Global Flags:
  -j, --json-output   If true, user-oriented output will be in JSON format.
```

But when you run it:
```
$ ddev share --basic-auth username:pass1234 
Error: unknown flag: --basic-auth
Usage:
  ddev share [project] [flags]
...
```

## How This PR Solves The Issue

Adds `--basic-auth` flag to the `ddev share` command.

## Manual Testing Instructions

```bash
ddev start
ddev share --basic-auth username:pass1234
```

## Related Issue Link(s)

* https://github.com/ddev/ddev/pull/3875

The same functionality can be achieved by running:
```bash
ddev config --ngrok-args '--basic-auth username:pass1234'
ddev share
```


<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/4719"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

rfay commented:
The docs failure is a result of move from buildkite.com/drud to buildkite.com/ddev. Not to worry.

rfay reviewed (changes requested):
It looks to me like the fundamental problem is that https://github.com/ddev/ddev/pull/3875 didn't make any sense. It dealt with the DDEV args instead of dealing with the ngrok args. 

It might make sense to revert #3875 first, although that's what made you think this should work I think right?

@tyler36 could you take another look at this?

stasadev commented:
I personally prefer `--ngrok-args`, so you don't depend on ngrok changes.

If no one has noticed in almost a year that `--basic-auth` doesn't work, let's just remove it.

rfay commented:
I agree, let's let @tyler36 weigh in and then we'll go forward. I'll need you to lead and test because I turned off the paid subscription after they hiked it so very much.  I definitely think we should be exploring use of cloudflare or other possibilities too.

stasadev commented:
Alright, in the meantime I prepared #4766. I don't have a paid ngrok account myself, but I assume it works, because it gives me a legit error:

```text
ERROR:  Custom subdomains are a feature on ngrok's paid plans.
ERROR:  Failed to bind the custom subdomain 'asd' for the account 'stasadev'.
ERROR:  This account is on the 'Free' plan.
ERROR:  
ERROR:  Upgrade to a paid plan at: https://dashboard.ngrok.com/billing/subscription
ERROR:  
ERROR:  ERR_NGROK_313
ERROR: 
```

tyler36 commented:
Works as intended.

## Test

### Current

1. Confirm current version.

    ```shell
    $ ddev -v
    ddev version v1.21.6
    ```

1. Start DDEV.

    ```shell
    $ ddev start
    ...
    Project can be reached at https://d10-base-demo.ddev.site https://127.0.0.1:54802
    ```

1. Attempt share.

    ```shell
    $ ddev share --basic-auth username:pass1234
    Error: unknown flag: --basic-auth
    ...
    ```

1. Stop project.

### This PR

1. Confirm current version.

    ```shell
    $ ddev -v
    ddev version v1.21.5-2-g9cb0ab0a
    ```

1. Attempt share.

    ```shell
    $ ddev share --basic-auth username:pass1234
    ...
    Account                       user13 (Plan: Free)
    Forwarding                    https://f591-124-37-96-190.ngrok.io -> http://127.0.0.1:54976
    ```

1. Confirmed site is accessible only with username/password provided.

### Global config

1. Confirm current version.

    ```shell
    $ ddev -v
    ddev version v1.21.6
    ```

1. Configure

    ```shell
    ddev config --ngrok-args '--basic-auth username:pass1234'
    ...
    Account                       user13 (Plan: Free)
    Forwarding                    https://3d3b-124-37-96-190.ngrok.io -> http://127.0.0.1:55097
    ```

1. Confirmed site is accessible only with username/password provided.

Works as intended. However, `~/.ddev/config.yaml` is typically managed by git which leads to authentication details potentially being committed. Giving "bad actors" insight into username/password combos is not good practice.
I think moving forward with this PR is the right way to proceed.

## Alternative

Add a NGROK helper command (`ddev ngrok`) which passes all arguments through (similar to `ddev composer`).
This would:

- Be future proof if they change their API
- Could be aliased to `ddev share` for backwards compatible
- Prepare DDEV for alternative sharing providers as/if required.

stasadev commented:
> .ddev/config.yaml is typically managed by git which leads to authentication details potentially being committed.

Yeah, that's a really good argument.

I rebased and added my [changes](https://github.com/ddev/ddev/pull/4719/commits/82131ce80310854479baf1cdbfe2c40b530c62c3) from another PR, as some of the command descriptions were out of date. I've also moved the required ngrok account information to the top because people tend to pay attention to what's written first.

stasadev commented:
@rfay it is ready, the only thing that has changed since @tyler36 review is the output of `ddev share --help` and the docs with @mattstein help.

If that's okay with you, let's merge :rocket: 

Alternative PR #4766 with revert should be closed.

rfay commented:
Thanks, will certainly get to it. Sorry I'm slow, backed up on many things and have been gone for several days. I don't want to merge until I manually test it.  Thanks so much for this!

rfay reviewed (approved):
This looks and works great, thanks for all the work on it!

rfay commented:
Looks like you already took a pass at this @mattstein but it may have changed since then.

mattstein reviewed (approved):
Looks good to me!
